js/src/reports/utils.test.js

```diff
@@ -0,0 +1,296 @@
+/**
+ * Internal dependencies
+ */
+import {
+	freeFields,
+	paidFields,
+	MISSING_FREE_LISTINGS_DATA,
+} from '.~/data/utils';
+import {
+	getIdsFromQuery,
+	aggregateIntervals,
+	sumToPerformance,
+	addBaseToPerformance,
+} from './utils';
+
+// Copied from https://github.com/woocommerce/woocommerce-admin/blob/b35156dcf17b44b3a81ea4b4528445b432917fd5/packages/navigation/src/test/index.js#L184-L221
+describe( 'getIdsFromQuery', () => {
+	it( 'if the given query is empty, should return an empty array', () => {
+		expect( getIdsFromQuery( '' ) ).toEqual( [] );
+	} );
+
+	it( 'if the given query is undefined, should return an empty array', () => {
+		expect( getIdsFromQuery( undefined ) ).toEqual( [] );
+	} );
+
+	it( 'if the given query is does not contain any coma-separated numbers, should return an empty array', () => {
+		expect( getIdsFromQuery( 'foo123,bar,baz1.' ) ).toEqual( [] );
+	} );
+
+	describe( 'if the given query contains numbers', () => {
+		it( 'should return an array of them', () => {
+			expect( getIdsFromQuery( '77,8,-1' ) ).toEqual( [ 77, 8, -1 ] );
+		} );
+		it( 'should consider `0` a valid id', () => {
+			expect( getIdsFromQuery( '0' ) ).toEqual( [ 0 ] );
+			expect( getIdsFromQuery( '77,0,1' ) ).toEqual( [ 77, 0, 1 ] );
+		} );
+		it( 'should map floats to integers', () => {
+			expect( getIdsFromQuery( '77,8.54' ) ).toEqual( [ 77, 8 ] );
+		} );
+		it( 'should ignore duplicates', () => {
+			expect( getIdsFromQuery( '77,8,8' ) ).toEqual( [ 77, 8 ] );
+			// Consider two floats that maps to the same integer a duplicate.
+			expect( getIdsFromQuery( '77,8.5,8.4' ) ).toEqual( [ 77, 8 ] );
+		} );
+		it( 'should ignore non numbers entries in the coma-separated list', () => {
+			expect( getIdsFromQuery( '77,,8,foo,null,9' ) ).toEqual( [
+				77,
+				8,
+				9,
+			] );
+		} );
+	} );
+} );
+
+describe( 'aggregateIntervals', () => {
+	function toIntervals( ...tuples ) {
+		return tuples.map( ( [ interval, ...values ] ) => ( {
+			interval,
+			subtotals: values.reduce(
+				( acc, value, i ) => ( {
+					...acc,
+					[ paidFields[ i ] ]: value,
+				} ),
+				{}
+			),
+		} ) );
+	}
+
+	it( 'if both `intervals` parameters are not provided, should return null', () => {
+		expect( aggregateIntervals() ).toBeNull();
+		expect( aggregateIntervals( null ) ).toBeNull();
+		expect( aggregateIntervals( null, null ) ).toBeNull();
+		expect( aggregateIntervals( undefined, null ) ).toBeNull();
+	} );
+
+	it( 'should sort aggregated intervals in ascending order of string code by `interval`', () => {
+		const intervals1 = toIntervals(
+			[ '2021-01' ],
+			[ '2021-03' ],
+			[ '2021-05' ],
+			[ '2021-07' ]
+		);
+		const intervals2 = toIntervals(
+			[ '2021-02' ],
+			[ '2021-04' ],
+			[ '2021-06' ],
+			[ '2021-08' ]
+		);
+		const expectedIntervals = toIntervals(
+			[ '2021-01' ],
+			[ '2021-02' ],
+			[ '2021-03' ],
+			[ '2021-04' ],
+			[ '2021-05' ],
+			[ '2021-06' ],
+			[ '2021-07' ],
+			[ '2021-08' ]
+		);
```

Comment:
I understand `expectedIntervals` is related to `intervals1` and `intervals2` so in a way it can make things easier to read by placing them together. We can live with this, don't worry too much about this. 😄

Reply:
Adjusted by https://github.com/woocommerce/google-listings-and-ads/pull/866/commits/d55dc0405f9f550a6390cedb5fdcea5fa1312063